// dboper/lib/dboper.js
/**
 * Created by Win 7 on 2014/8/18.
 */
var mongodb = require('../db/msession');
var ObjectID = require('mongodb').ObjectID;
/**
 * 根据id查询指定表
 * @param table
 * @param id
 * @param callback
 */
exports.findbyid=function(table,id,callback){
    mongodb.close();
    mongodb.open(function(err, db) {
        if (err) {
            return callback(err);
        }

        db.collection(table, function(err, collection) {
            if (err) {
                mongodb.close();
                return callback(err);
            }

            collection.findOne({_id:new ObjectID(id)}, function(err, doc) {

                mongodb.close();
                if (doc) {


                    console.log(doc._id+"-=-=-=-")
                    callback(err, doc);
                } else {
                    callback(err, null);
                }
            });
        });
    });

}
/**
 * 根据指定条件查询，返回结果按id排序
 * @param table
 * @param query
 * @param callback
 */
exports.findbyquery=function(table,query,callback){
    mongodb.close();
    mongodb.open(function(err,db){
        if(err){
            return callback(err);
        }
        db.collection(table,function(err,collection){
            if(err){
                mongodb.close();
                return callback(err);
            }

            collection.find(query).sort({_id:-1}).toArray(function(err,doc){
                if(err){
                    callback(err,null);
                }


                callback(null,doc);
            });
        });
    });
}
/**
 * 根据日期倒序排查询所有结果
 * @param table
 * @param callback
 */
exports.getAll=function(table,callback){
    mongodb.close();
    mongodb.open(function(err,db){
        if(err){
            return callback(err);
        }
        db.collection(table,function(err,collection){
            if(err){
                mongodb.close();
                return callback(err);
            }
            var query={};

            collection.find().sort({createdate:-1}).toArray(function(err,docs){

                mongodb.close();
                if(err){
                    callback(err,null);
                }


                callback(null,docs);

            });
        });
    });
}
/**
 * 根据id删除记录
 * @param table
 * @param id
 * @param callback
 */
exports.deletebyid=function(table,id,callback){
    mongodb.close();
    mongodb.open(function(err,db){
        if(err){
            return callback(err);
        }
        db.collection(table,function(err,collection){
            if(err){
                mongodb.close();
                return callback(err);
            }
            collection.remove({_id:new ObjectID(id)},  {safe:true},function(err,result){

                mongodb.close();
                callback(err,result);

            });
        });
    });
}
/**
 * 根据用户名密码查询用户表
 * @param table
 * @param username
 * @param userpasswod
 * @param callback
 */
exports.getbyname=function(table,username,userpasswod,callback){
    mongodb.close();
    mongodb.open(function(err, db) {
        if (err) {
            return callback(err);
        }

        db.collection(table, function(err, collection) {
            if (err) {
                mongodb.close();
                return callback(err);
            }
            collection.findOne({username:username,Logpwd:userpasswod}, function(err, doc) {

                mongodb.close();
                if (doc) {


                    callback(err, doc);
                } else {
                    callback(err, null);
                }
            });
        });
    });
}
/**
 * 根据id更新指定表
 * @param table
 * @param obj
 * @param id
 * @param callback
 */
exports.update=function(table,obj,id,callback){
    mongodb.close();
    mongodb.open(function(err,db){
        if(err){
            return callback(err);
        }
        db.collection(table,function(err,collection){
            if(err){
                mongodb.close();
                return callback(err);
            }

            collection.update({_id:new ObjectID(id)},
                {$set:obj},
                {safe:true},function(err,result){
                    mongodb.close();
                    callback(err,result);

                });
        });
    });
}
/**
 * 向指定表添加数据
 * @param table
 * @param obj
 * @param callback
 */
exports.save=function(table,obj,callback){
    mongodb.close();
    mongodb.open(function(err,db){
        if(err){

            return callback(err);
        }
        db.collection(table,function(err,collection){
            if(err){
                mongodb.close();
                return callback(err);
            }
            collection.insert(obj,{safe:true},function(err,result){
                mongodb.close();
                callback(err,result);
            });
        });
    })
}
